clippy_lints/src/question_mark.rs

```diff
@@ -271,6 +275,139 @@ fn check_is_none_or_err_and_early_return<'tcx>(cx: &LateContext<'tcx>, expr: &Ex
+    let arm_body = arm.body.peel_blocks().peel_drop_temps();
+    let happy_ctor = match mode {
+        TryMode::Result => ResultOk,
+        TryMode::Option => OptionSome,
+    };
+
+    // Check for `Ok(val)` or `Some(val)`
+    if arm.guard.is_none()
+        && let Some(val_binding) = extract_ctor_call(cx, happy_ctor, arm.pat)
+        // Extract out `val`
+        && let Some(binding) = extract_binding_pat(val_binding)
+        // Check body is just `=> val`
+        && let Res::Local(body_local_ref) = path_res(cx, arm_body)
+        && binding == body_local_ref
+    {
+        true
+    } else {
+        false
+    }
+}
+
+// Check `arm` is `Err(err) => return Err(err)` or `None => return None`.
+fn check_arm_is_sad<'tcx>(cx: &LateContext<'tcx>, mode: TryMode, arm: &Arm<'tcx>) -> bool {
+    if arm.guard.is_some() {
+        return false;
+    }
+
+    let arm_body = arm.body.peel_blocks().peel_drop_temps();
+    match mode {
+        TryMode::Result => {
+            // Check that pat is Err(val)
+            if let Some(ok_pat) = extract_ctor_call(cx, ResultErr, arm.pat)
+                && let Some(ok_val) = extract_binding_pat(ok_pat)
+                // check `=> return Err(...)`
+                && let ExprKind::Ret(Some(wrapped_ret_expr)) = arm_body.kind
+                && let ExprKind::Call(ok_ctor, [ret_expr]) = wrapped_ret_expr.kind
+                && is_res_lang_ctor(cx, path_res(cx, ok_ctor), ResultErr)
+                // check `...` is `val` from binding
+                && let Res::Local(ret_local_ref) = path_res(cx, ret_expr)
+                && ok_val == ret_local_ref
+            {
+                true
+            } else {
+                false
+            }
+        },
+        TryMode::Option => {
+            // Check the pat is `None`
+            if is_res_lang_ctor(cx, path_res(cx, arm.pat), OptionNone)
+                // Check `=> return None`
+                && let ExprKind::Ret(Some(ret_expr)) = arm_body.kind
+                && is_res_lang_ctor(cx, path_res(cx, ret_expr), OptionNone)
+            {
+                true
+            } else {
+                false
+            }
+        },
+    }
+}
+
+fn check_arms_are_try<'tcx>(cx: &LateContext<'tcx>, mode: TryMode, arm1: &Arm<'tcx>, arm2: &Arm<'tcx>) -> bool {
+    (check_arm_is_happy(cx, mode, arm1) && check_arm_is_sad(cx, mode, arm2))
+        || (check_arm_is_happy(cx, mode, arm2) && check_arm_is_sad(cx, mode, arm1))
+}
+
+fn check_if_try_match<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+    if let ExprKind::Match(scrutinee, [arm1, arm2], MatchSource::Normal | MatchSource::Postfix) = expr.kind
+        && !expr.span.from_expansion()
+        && let Some(mode) = find_try_mode(cx, scrutinee)
+        && check_arms_are_try(cx, mode, arm1, arm2)
+    {
+        let mut applicability = Applicability::MachineApplicable;
+        let mut snippet = snippet_with_applicability(cx, scrutinee.span, "...", &mut applicability).into_owned();
```

Comment:
This also needs to use either `snippet_with_context` with the match expression's SyntaxContext or pass `scrutinee.span.source_callsite()` so we don't take snippets in macro expansions:
```rs
macro_rules! opt {
    () => { None };
}
let v = match opt!() {
    Some(x) => x,
    None => return None,
};
```
This should suggest `opt!()?;` instead of `None?;` (would be good to have this as a a test).

Reply:
Done both.